reviewrot/pagurestack.py

```diff
@@ -155,16 +155,40 @@ def get_last_comment(self, res):
                 body=str(comments[-1]['comment']),
                 created_at=last_comment_date)
 
-    @staticmethod
-    def _avatar(username):
-        """ Return the avatar of a given pagure user.
+    def _avatar(self, username, ssl_verify=True):
+        """ Return the avatar URL of a given pagure user.
 
-        Pagure avatars have a predictable URL structure.
+        This method will first try to fetch the avatar URL from pagure API.
+        However, this information has only recently been exposed. As a fallback,
+        this method will construct the avatar URL based on the username. This
+        provides an avatar, but it's not the user's avatar since that one is
+        based on user's email address which is not publicly exposed.
         """
-        query = urllib.parse.urlencode({'s': 64, 'd': 'retro'})
-        openid = u'http://%s.id.fedoraproject.org/' % username
-        idx = hashlib.sha256(openid.encode('utf-8')).hexdigest()
-        return "https://seccdn.libravatar.org/avatar/%s?%s" % (idx, query)
+        request_url = "{}/api/0/user/{}".format(self.instance, username)
+        log.debug("Looking for avatar URL for user %s", username)
+        try:
+            response = self._call_api(url=request_url, ssl_verify=ssl_verify)
+        except requests.exceptions.HTTPError:
+            raise ValueError("User {} not found!".format(username))
+
+        avatar_query = {'s': 64, 'd': 'retro'}
+        avatar_url = response.get('user', {}).get('avatar_url')
+        if avatar_url:
+            # Make sure we're getting the resolution we expect.
+            url_parts = urllib.parse.urlparse(avatar_url)
+            url_query = urllib.parse.parse_qs(url_parts.query)
+            url_query.update(avatar_query)
+            url_query = urllib.parse.urlencode(url_query)
```

Comment:
If the value in the dictionary is not a string, urlencode will convert into a string, for instance `urllib.parse.urlencode({'s': ['16'], 'd': ['retro']})` will generate `'s=%5B%2716%27%5D&d=%5B%27retro%27%5D'`. Since `urllib.parse.parse_qs` has the dictionary values as lists, if the `url_query.update(avatar_query)` doesn't update all the values in the dictionary, you won't get the result you want.

Reply:
`url_query` is derived from the query parameters of the parsed URL, for example `'s=16&d=retro'`. In what scenario will a query parameter be parsed into something like `{'s': ['16'], 'd': ['retro']}` ?

Reply:
@lcarva 
```python
>>> avatar_url = 'https://seccdn.libravatar.org/avatar/ecae5ad9b37619a38e699de00049e449f120b433c9f26a8beb05b990dfc54c9d?s=16&d=retro'
>>> url_parts = urllib.parse.urlparse(avatar_url)
>>> url_query = urllib.parse.parse_qs(url_parts.query)
>>> url_query = urllib.parse.urlencode(url_query)
>>> print(url_query)
s=%5B%2716%27%5D&d=%5B%27retro%27%5
```

Reply:
Aha! Good catch! I'll fix it up.

Reply:
Fixed it with this: `urlencode(..., doseq=True)`